pkg/operators/operators.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operators
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type GadgetContext interface {
+	ID() string
+	Context() context.Context
+	GadgetDesc() gadgets.GadgetDesc
+	Logger() logger.Logger
+}
+
+type (
+	EnricherFunc func(any) error
+)
+
+var ErrAsync = errors.New("async")
+
+type Operator interface {
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// GlobalParamDescs will return global params (required) for this operator
+	GlobalParamDescs() params.ParamDescs
+
+	// ParamDescs will return params (required) per gadget instance of the operator
+	ParamDescs() params.ParamDescs
+
+	// Dependencies can list other operators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.GadgetDesc) bool
+
+	// Init allows the operator to initialize itself
+	Init(params *params.Params) error
+
+	// Close allows the operator to clean up stuff prior to exiting
+	Close() error
+
+	// Instantiate is called before a gadget is run with this operator.
+	// This must return something that implements OperatorInstance.
+	// This is useful to create a context for an operator by wrapping it.
+	// Params given here are the ones returned by ParamDescs()
+	Instantiate(gadgetContext GadgetContext, gadgetInstance any, params *params.Params) (OperatorInstance, error)
+}
+
+type OperatorInstance interface {
+	// Name returns the name of the operator instance
+	Name() string
+
+	// PreGadgetRun in called before a gadget is run
+	PreGadgetRun() error
+
+	// PostGadgetRun is called after a gadget is run
+	PostGadgetRun() error
```

Comment:
I just renamed the `Close()` on `OperatorInstances` back to `PostGadgetRun()` to avoid confusion with the `Operator` interface. So we have `Init()` and `Close()` on the Operator, but `PreGadgetRun()` and `PostGadgetRun()` on the `OperatorInstance` - and I think it really is less confusing this way.